Zombienet-SDK test for the happy path of the new staking-async flow

tdimitrov opened the pull request:
Adds a zombienet test for the happy case of new staking flow where staking-async pallet is deployed on AH.

The test uses test/fake runtimes for RC and AH-Next. It covers the following scenario:
- AH and RC networks are started
- Election is started on AH
- Election yields the expected number of validators
- The new active set is delivered to RC
- On session change the active set on RC is changed and AH is notified

The test takes quite a lot of time (~11 minutes) to execute and requires the chainspecs for fake RC and AH-Next to be prebuilt in the root directory of the test module.

tdimitrov commented on substrate/zombienet-sdk/tests/staking_next/tests.rs:
```diff
@@ -0,0 +1,380 @@
+fn parse_session_report_received(fields: Composite<u32>) -> Option<RcClientEvent> {
+	if let subxt::ext::scale_value::Composite::Named(fields) = fields {
+		assert!(fields.len() == 4, "Expected 4 fields");
+		assert!(fields[1].0 == "activation_timestamp", "Expected activation_timestamp");
+
+		if let ValueDef::Variant(activation_timestamp_variant) = &fields[1].1.value {
+			if activation_timestamp_variant.name == "None" {
+				return Some(RcClientEvent::SessionReportReceived { activation_time_stamp: None })
+			}
+
+			assert!(activation_timestamp_variant.name == "Some", "Expected Some");
+
+			if let subxt::ext::scale_value::Composite::Unnamed(act) =
+				&activation_timestamp_variant.values
+			{
+				if let ValueDef::Composite(a) = &act[0].value {
+					if let subxt::ext::scale_value::Composite::Unnamed(a) = &a {
+						return Some(RcClientEvent::SessionReportReceived {
+							activation_time_stamp: Some(a[0].as_u128().unwrap()),
+						})
+					} else {
+						log::info!("act: {:?}", act);
+						assert!(false, "Expected ValueDef::Primitive`(");
+					}
+				} else {
+					log::info!("act: {:?}", act);
+					assert!(false, "Expected ValueDef::Composite(");
+				}
+			} else {
+				assert!(false, "Expected unnamed composite");
+			}
+		} else {
+			assert!(false, "Expected variant");
+		}
+	} else {
+		assert!(false, "Expected named composite");
+	}
+
+	None
+}
```
@jsdw could you please have a look at these functions? Is there a better way to do them?

tdimitrov replied:
As discussed offline, this will probably easier with the static api. I'll give it a shot.

jsdw replied:
Do let me know if you run into issues with that approach :)

kianenigma commented on substrate/zombienet-sdk/tests/lib.rs:
```diff
@@ -0,0 +1,4 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+mod staking_next;
```
Why have you chosen to have these tests to be in `./substrate` vs in `./substrate/frame/staking-async`? 

Also noting the new name is staking-async, not next.

tdimitrov replied:
It's a common pattern to have all zombienet tests in `polkadot/zombiened-sdk-tests`, `substrate/zombienet-sdk-tests`, etc. 

But I've got no strong preference on this. If you want, I'll move them to to `./substrate/frame/staking-async`.

I'll fix the name.

tdimitrov replied:
Both are fixed.

kianenigma commented on substrate/zombienet-sdk/tests/staking_next/mod.rs:
```diff
@@ -0,0 +1,82 @@
+
+mod tests;
+
+/// Sets `ValidatorCount` in staking pallet to 500.
```
```suggestion
/// Sets `ValidatorCount` in staking pallet to `validator_count`.
```

kianenigma replied:
Also worth adding a comment that this is due to some mysterious bug, otherwise it should not be used/needed.

Has there been any updates on that? do we have any new leads as to why it is needed?

tdimitrov replied:
I've added the comment. I still haven't investigated why the value from the chainspec doesn't get set to the storage item.

kianenigma commented on substrate/zombienet-sdk/tests/staking_next/mod.rs:
```diff
@@ -0,0 +1,82 @@
+) -> Result<(), anyhow::Error> {
+	let set_validator_count = dynamic::tx("Staking", "set_validator_count", vec![validator_count]);
+	let sudo = dynamic::tx("Sudo", "sudo", vec![set_validator_count.into_value()]);
+	let alice = dev::alice();
```
This will work fine, but to be more correct you should query `Sudo::key` here, see who it is, and if it is one of the dev accounts (alice, bob etc.), sign and send it, and if not fail.

In other words, you are assuming alice is sudo, which may or may not be.

kianenigma commented on substrate/zombienet-sdk/tests/staking_next/mod.rs:
```diff
@@ -0,0 +1,82 @@
+		})
+		.with_parachain(|p| {
+			p.with_id(1100)
+			.with_chain_spec_path("parachain.json") // TODO: how to autogenerate this?
```
esp if you move it to `./substrate/frame/staking-next`, I think it is okay for now to say in the README.md:

please first run `build-and-run-zn.sh compile`, and tweak this bash script to have an option to just prepare the two chain-specs, and not run ZN.

kianenigma commented on substrate/zombienet-sdk/tests/staking_next/tests.rs:
```diff
@@ -0,0 +1,380 @@
+	ext::scale_value::{Composite, ValueDef},
+	OnlineClient, PolkadotConfig,
+};
+
```
```suggestion
```

Would allow rustfmt to organize them automatically ;)

kianenigma commented on substrate/zombienet-sdk/tests/staking_next/tests.rs:
```diff
@@ -0,0 +1,380 @@
+
+// Events generated by rc_client. Event here doesn't necessary mean 'event in the substrate
+// context'. It can be anything representing a point of interest for the tests.
+enum RcClientEvent {
```
I think subxt has a cli which allows you to generate a metadata from the wasm file, and within that you have access to the subxt-compliant version of all of the runtime types, including the concrete event types. So you could, instead of this, assert that some events are emitted.

tdimitrov replied:
That's true but in this case this event is not a runtime event.
It's an 'event' such as 'something happened in the test which we want to handle and probably change state. In this test there is 100% match between runtime event in test event but in general this might not be the case.

I should pick a better name since this causes confusion.

kianenigma commented on substrate/zombienet-sdk/tests/staking_next/tests.rs:
```diff
@@ -0,0 +1,380 @@
+	Ok(())
+}
+
+// TODO: There should be a better way to do this.
```
my guess would again be subxt typegen here.

tdimitrov replied:
Here I agree 100% - I'm reworking this right now.

kianenigma reviewed (commented):
Would be ideal to simplify the manual-data parsing here, before we extend it to more test scenarios. 

Strongly in favor of moving this to `staking-async` crate, under a new sub-crate over there to keep it more isolated.

tdimitrov commented:
> Would be ideal to simplify the manual-data parsing here, before we extend it to more test scenarios.

The static interface is problematic for CI since we don't want to commit the metadata and at the same time we need it generated somehow. To avoid complicated build scripts I am sticking with the dynamic subxt interface but with some advise from @jsdw  I reworked the parsing. I think it's way better now.

> Strongly in favor of moving this to `staking-async` crate, under a new sub-crate over there to keep it more isolated.

I've also moved the test in `staking-async`, as an ignored integration test. That way CI won't pick it up but we still can run it manually.

jsdw commented:
> The static interface is problematic for CI since we don't want to commit the metadata and at the same time we need it generated somehow.

In case it helps, Subxt does this by spinning up a Substrate node in a build.rs file, downloading metadata from it and then providing an instance of the generated interface from it here: 

https://github.com/paritytech/subxt/blob/master/testing/test-runtime/build.rs

(If the node is already spun up at the time the tests start running, you can also generate an interface given a URL to an RPC node ie like this: https://docs.rs/subxt/latest/subxt/attr.subxt.html#connecting-to-a-node-to-download-metadata-via-runtime_metadata_insecure_url--)

But yeah, I appreciate that this adds a little complexity. The static interface def makes life easier though :)

kianenigma commented on substrate/frame/staking-async/Cargo.toml:
```diff
@@ -35,18 +35,26 @@ sp-staking = { features = ["serde"], workspace = true }
 sp-npos-elections = { workspace = true, default-features = true }
 sp-tracing = { workspace = true, default-features = true }
 substrate-test-utils = { workspace = true }
+subxt = { workspace = true }
```
Let's move them to a separate crate `staking-async/zn-test` to not increase the compile time of the rust unit tests?
